Frameworks/CoreGraphics/CGContext.mm

```diff
@@ -1913,71 +1943,155 @@ void CGContextSetPatternPhase(CGContextRef context, CGSize phase) {
     // Undo assumed inversion about Y axis
     CGAffineTransform textTransform = CGAffineTransformScale(textMatrix, 1.0, -1.0);
     CGAffineTransform deviceTransform = CGContextGetUserSpaceToDeviceSpaceTransform(this);
+    CGAffineTransform combinedTransform = CGAffineTransformConcat(textTransform, deviceTransform);
 
     // Snap the vertical baseline to the nearest pixel to avoid blurring on devices with vertical antialiasing
     textTransform.ty = std::round(textTransform.ty);
 
-    CGAffineTransform finalTextTransform = CGAffineTransformConcat(textTransform, deviceTransform);
-    if ((fabs(finalTextTransform.a * glyphRun->fontEmSize) <= c_glyphThreshold &&
-         fabs(finalTextTransform.b * glyphRun->fontEmSize) <= c_glyphThreshold) ||
-        (fabs(finalTextTransform.d * glyphRun->fontEmSize) <= c_glyphThreshold &&
-         fabs(finalTextTransform.c * glyphRun->fontEmSize) <= c_glyphThreshold)) {
-        TraceWarning(TAG, L"Glyphs too small to be rendered");
-
-        // Not a failure state! Not drawing the glyphs is *okay*.
-        return S_FALSE;
+    bool fastPathTextDrawing =
```

Comment:
didn't you have more granularity in iteration 1 than this? (can you fast-path some stuff even if textDrawingMode isn't Fill?)